airbyte-integrations/connectors/source-googleanalytics-singer/source_googleanalytics_singer/source.py

```diff
@@ -44,14 +47,38 @@ def configure(self, raw_config: json, temp_dir: str) -> json:
         with open(credentials, "w") as fh:
             fh.write(raw_config["credentials_json"])
         raw_config["key_file_location"] = credentials
+        if raw_config.get("custom_reports", "").strip() and json.loads(raw_config["custom_reports"]):
+            custom_reports_data = json.loads(raw_config["custom_reports"])
+            self._validate_custom_reports(custom_reports_data)
+            raw_config["reports"] = self._get_reports_file_path(temp_dir, custom_reports_data)
         return super().configure(raw_config, temp_dir)
 
+    def _validate_custom_reports(self, custom_reports_data: List[dict]):
+        custom_reports_schema = json.loads(pkgutil.get_data("source_googleanalytics_singer", "custom_reports_schema.json"))
+        if not Draft4Validator(custom_reports_schema).is_valid(custom_reports_data):
+            error_messages = []
+            for error in Draft4Validator(custom_reports_schema).iter_errors(custom_reports_data):
+                error_messages.append(error.message)
+            raise Exception("An error occurred during custom_reports data validation:\n" + ";\n".join(error_messages))
```

Comment:
I think throwing an exception here will make methods like `check` fail entirely (not send a "check failed" message) because this method will get called before them, which is not in line with the specification. We'll unfortunately need to do this validation in the check method, and transform the config in each of the `check`, `discover` and `read` methods manually. 

To see what I mean, try changing your local `config.json` to contain an invalid JSON in the `custom_reports` field and run `python main_dev check --config secrets/json` and you'll see the command fails entirely (it should produce a `status:FAILED` message instead to indicate credentials are invalid).

Reply:
This will only be needed until the UI change is made to support inputting well formatted JSON in the UI